BMP Image Regression Tool

ElliotFreebairn opened the pull request:
The code/program helps the visual regression testing between authoritative Microsoft Office PDF's and imported Collabora Online PDF's. It converts each PDF page into a grey scale RGBA BMP, then performs pixel-level comparisons, and generates visual differences into another BMP.

### Usage
`./generate-bmp-diffs.sh <pdf_directories> <input_directory> <output_directory> [max_pages] [enable_minor_differences]`

**Arguments:**
- `pdf_directories`: Path to PDF's. Must include **both** authoritative and import files. Accepts individual files or nested directories. e.g. `./image-dump/some-folder/*.doc_authoritative.pdf` and `./image-dump/some-folder/*.doc_import.pdf`
- `input_directory`: Output directory for the converted BMP's. e.g. `input/some-folder/auth-page-1.bmp`
- `output_directory`: Output directory for image diffs. e.g. `output/some-folder/diff-page-1.bmp`
- `max_pages` (optional): Limits the number of pages converted per PDF
- `enable_minor_differences` (optional): Set to `true` to highlight minor pixel differences near edges in yellow. The default is `false`

**General overview of the image processing**

1. Convert PDFS to grey scale RGBA BMP's
2. Run Sobel edge detection 
3. Apply a blur mask with a given radius to suppress false positives near edges
4. For each pixel: Compare pixel values across images, mark differences in red, if above threshold(will implement the average threshold Michael talked about soon). Optionally if `enable_minor_differences` is true, mark the minor edge differences in yellow (with stricter threshold)

Still lots of work to do. Main thing right now is to make sure that it can run on windows. Need to check if Poppler and Imagemagick can run on the windows machine and the shell scripts run fine. The c++ code should be fine as I'm using standard libraries.

**General example of script:**
`./generate-bmp-diffs.sh ./image-dump/ ./input/ ./output/  5 true`

ElliotFreebairn committed:
First commit: convert PDF's to BMP's and produce a diff

A shell script converts the given PDF's to BMP's.
They are then read in via the c++ program and output BMP is produced

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Removing bmp's and updating .gitignore

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Pdf-to-bmp script can convert pdf's from an image-dump

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Add a check to see if PDF's are already converted

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Make sure image-dump is passed and new diff script

The new diff script goes through each input dir and compares two BMP's.
A BMP is produced and stored in an output dir

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Add edge detection to deal with text quirks

Sobel edge detection will help blur out pixels around text.
This helps discard "differences" rooting from anti-aliasing effects.

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Create blurred edge mask from pair of BMP's

The pair of masks are intersected and areas where both images have
edges and then ignored when highlighting differences.

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Refactor: create header files

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Removing scripts folder

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Tweaking pixel comparison thresholds

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
PDF's can be located in sub-directories or not

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Added option to highlighting minor differences

These minor differences can also be thought of as false positives
This is because they are located near edges/text, which most of the time
is caused by anti-aliasing/rendering quirks.

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Refactor: convert to snake_case

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Iterate over all pages(BMP's) and create diff's

Example files: auth-page1, auth-page2, import-page1, import-page2
The program will take an auth page and its corresponding import page and produce
a diff for that page.

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Logging stats when bmp's diff is finished

red and yellow count are tracked for the authoritative image
and the counts + percentages compared to total pixels are printed
at the end of run of the program

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Fix: minor differences option made pixels near edges red

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Improved the shell scripts.

The generate-bmp-diffs.sh is now more versatile. The user has to enter the directory the pdf lives.
They are also able to set where the output off the diffs will be.
They also can set some options like max pages and enable minor differences.
This means that the user no longer has to use pdf-to-bmp.sh has the current script handles that, with the given arguments.

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Bug fixing, comments, and MPLv2 headers

One bug was that if the bmp was larger than the other it would do funny things.
Changed the indexing for bmps's so they are custom to each other, not based off
min_width. Cleaned some the pixel code up with.

Added case checking into main.

Adding comments to the files which should be ignored by git.

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

pedropintosilva commented:
It seems to have conflicts. Vivek as Elliot's mentor could you please look at the code? @thebearon could you pretty please test it and give your opinion? Adding @timar in case he has anything to add.

ElliotFreebairn commented:
> It seems to have conflicts. Vivek as Elliot's mentor could you please look at the code? @thebearon could you pretty please test it and give your opinion? Adding @timar in case he has anything to add.

I'm fixing the conflicts now

mmeeks reviewed (changes requested):
Good to read the code can you:

1. add license MPLv2 headers to all your new files in an appropriate form
2. remove the stb_image stuff with all the extraneous crediting requirements. Do we really need it ? surely loading and saving a BMP without that is somewhat trivial. It seems the PDF conversion already generates PDF renders as PNG - so we should just load PNGs instead I think: they are far smaller. If you can find some MPLv2 code that doesn't drag around so much junk =) we could use that instead of course.
3. Are we doing a single BMP page compare and then aggregating stats in shell ? in general the more shell we write the worse life is in the long run =) can we not process multiple pages in the C++ itself and aggregate statistics for the whole file ?

Justin wrote a chunk of IIRC Python to handle the pixel comparison IIRC - I would expect to see some changes to that to hook your tool in instead of ImageMagic, rather than an entirely new shell-script thing ? =) or did I miss something.

anyhow - in general looks promising :-) I'd love to see the output.

Thanks!

ElliotFreebairn committed:
Got rid of the dangerous reinterpret cast and replaced with safer casting.

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Stop converting to PNG's, convert straight to BMP

Also removing some whitespace in BMP.cpp

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn commented:
> Good to read the code can you:
> 
> 1. add license MPLv2 headers to all your new files in an appropriate form
> 2. remove the stb_image stuff with all the extraneous crediting requirements. Do we really need it ? surely loading and saving a BMP without that is somewhat trivial. It seems the PDF conversion already generates PDF renders as PNG - so we should just load PNGs instead I think: they are far smaller. If you can find some MPLv2 code that doesn't drag around so much junk =) we could use that instead of course.
> 3. Are we doing a single BMP page compare and then aggregating stats in shell ? in general the more shell we write the worse life is in the long run =) can we not process multiple pages in the C++ itself and aggregate statistics for the whole file ?
> 
> Justin wrote a chunk of IIRC Python to handle the pixel comparison IIRC - I would expect to see some changes to that to hook your tool in instead of ImageMagic, rather than an entirely new shell-script thing ? =) or did I miss something.
> 
> anyhow - in general looks promising :-) I'd love to see the output.
> 
> Thanks!

@mmeeks Thank you for the comment, I'm working through these changes as we speak. I've created a small BMP reader and writer to replace stb_image, but I'm a little confused on what you mean in point 2. Are you talking about the PDF conversion done via mso-test? And when you mean by 'Load PNG', does that mean work with the PNG instead of BMP, or convert from PNG to BMP and skip the PDF? Sorry for the hassle

mmeeks commented:
Oh - if you have your own BMP reader and it works - lets just get going with that :-) as you say BMP is extremely easy to handle as a standalone format lets get that in and ignore the PNG reading change if that fits.

ElliotFreebairn committed:
Stop logging stats after each comparison, now at EOC

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Hooking pixelbasher into python script.

Its not working both on linux and windows. However, i need to replicate all the funtionality that justin implemented with image magic with the c++.

I'm a bit stuck on how to aggeragte the stats as i can't access values inside of the c++ objects unlike the imagemagick objects used .

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Replacing python imagemagick code with the c++ program

Changing the arguments passed to c++ so that it reflects what was passed previously in the python script.

Added the ability to compare round tripped MSO pdfs.

Fixing some referencing and memory bugs.

Starting to work on moving the logging from shell to csv files, like justin does in the script.

Still alot more work to do, but nearly there.

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Stats are now written to a csv file.

The values used and order of statistics written to the csv asre exactly the same as justin's python version.

Also fixed some minor bugs with wrong variable names and also decremented the blur radius to make the difference highllighting a bit more sensitive.

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Tweaking python script and c++ program to work together

Now you can run ./diff-pdf-page-statistics.py as you would before and it's producing nearly identical content (except from the better diffing bmp's)

Need to add the comparing previous versions to the c++ prgoram and check this works on windows nicely.

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn commented:
Don't worry about the force-pushing. I'm trying to figure how to remove some stuff from the commit history. Shoudn't take too long for me to figure out.

ElliotFreebairn commented:
> Don't worry about the force-pushing. I'm trying to figure how to remove some stuff from the commit history. Shoudn't take too long for me to figure out.

I'm getting some weird bugs trying to filter out those commits, so I've just forced-push my local repo with the current changes in it with the commit history still there.

In the meantime I'm going to fix these issues.

mmeeks commented:
git rebase -i <commit-before-you-started>

is your friend =) read the man page on it and have a play; you can't loose anything, and 'git reflog' allows you to git reset --hard <hash> your branch back to somewhere if you get lost ;-)

ElliotFreebairn commented:
> git rebase -i
> 
> is your friend =) read the man page on it and have a play; you can't loose anything, and 'git reflog' allows you to git reset --hard your branch back to somewhere if you get lost ;-)

Thank you, I've succesfully removed those commits, but it's including all the previous commits from the repo when i push. Is this is a bad thing?

ElliotFreebairn commented:
I'm having alot of trouble with the rebasing for some reason. It was causing conflicts and when I pushed it included previous commits from master. I'm going to take a break from trying to clear the history (until I can get some help) and work on some on the features you mentioned.

codewithvk commented:
@ElliotFreebairn When you're trying to remove a commit that's not the latest one, and it's depended on by later commits, you'll likely encounter merge conflicts during the rebase. These conflicts need to be resolved manually.

Which specific commit are you trying to remove?

ElliotFreebairn commented:
> @ElliotFreebairn When you're trying to remove a commit that's not the latest one, and it's depended on by later commits, you'll likely encounter merge conflicts during the rebase. These conflicts need to be resolved manually.
> 
> Which specific commit are you trying to remove?

@codewithvk  Hi, don't worry it's all good now. Thank you

ElliotFreebairn committed:
Refactoring argument passing logic

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Implementing regression checks between CO/LO versions.

The process the current LO version is diffed with the MSO original. Then the previous LO version is diffed with the MSO.

If a previous LO version pixel is correct and current version is wrong, the diff is coloured RED. If the previous is wrong and current is right, the diff is coloured GREEN. If both are wrong, the diff is coloured BLUE.

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Adding diffing for both LO  and MSO roundtripped files.

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Renaming variables to be more consistent and mirror the python scirpt.

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Fixed bug, where ms_conv_previous were passed and not lo_previous

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Removing the unused shell script files.

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

mmeeks commented:
This looks fine to me to merge - but I've not had time to go over the code & help that improve structurally.

ElliotFreebairn commented:
> This looks fine to me to merge - but I've not had time to go over the code & help that improve structurally.

No worries, maybe @codewithvk can help with that?

ElliotFreebairn commented:
I'm going to spend the rest of the day testing and refactoring the code before @codewithvk reviews it, hopefully making it easier for him.

codewithvk commented:
> I'm going to spend the rest of the day testing and refactoring the code before @codewithvk reviews it, hopefully making it easier for him.

@ElliotFreebairn  No problem, take your time.

One general suggestion, not specific to this project but applicable to any codebase: try to keep your commit messages concise, and use the description field to provide more detailed context. This helps others quickly understand the intent and background of each change.

Personally, I follow this approach for writing commit messages:
https://www.freecodecamp.org/news/how-to-write-better-git-commit-messages/
But feel free to use any standard that works best for you.

ElliotFreebairn committed:
Refactoring: cleaned up main.cpp to be more readable

Also some minor refactoring in other files and clearer error messages.

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Improving modularity and performance.

Split up the functions for comparing BMP's and edge detection to smaller and simpler functions.

Moved edge detection/creation of blurred edge mask to BMP, so It's only calculated once compared to each compare call.

Fixing instances where data was copied instead of referenced, reducing the performance.

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Refactor:  formatting inconsistent space/tab sizes.

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Feat: side by side image for visual comparison.

If the option
--image_dump is added, then for every page, an image is created with the diff, next to ms_original and LO/MS_CONV page.

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Refactor: improving comments and error messages

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

stbergmann commented on src/bmp.hpp:
```diff
@@ -0,0 +1,107 @@
+#include <cstddef>
+#include <vector>
+#include <iostream>
+#include <array>
```
here and elsewhere: if there's no better natural sort order for a block of similar things, it's best to just sort them lexicographically

stbergmann commented on src/bmp.hpp:
```diff
@@ -0,0 +1,107 @@
+
+	const BMPFileHeader &get_file_header() { return file_header; }
+	const BMPInfoHeader &get_info_header() { return info_header; }
+	const BMPColourHeader &get_colour_header() { return colour_header; }
```
the above three trivial getter functions are only used internally, so they could be private (and also `const`), and to internally access any data members, you don't actually need any getter functions, so, even better, they could be dropped completely

stbergmann commented on src/bmp.hpp:
```diff
@@ -0,0 +1,107 @@
+	int get_average_colour() const;
+	int get_non_background_pixel_count(int background_value) const;
+
+	std::vector<bool> sobel_edges(const BMP &image, int threshold = 40);
```
* instead of passing `*this` as an explicit `image` parameter, the function already has that `BMP` object implicitly available via `this` anyway (so no need for that `image` parameter; and the member function can be `const`)

* there's a single caller of this function, and it uses the default `threshold` argument, so you could just as well trun `threshold` into a (`const` or even `constexpr`) variable internally to that function

stbergmann commented on src/bmp.hpp:
```diff
@@ -0,0 +1,107 @@
+	BMPInfoHeader info_header;
+	BMPColourHeader colour_header;
+	std::vector<uint8_t> data;
+	std::vector<bool> blurred_edge_mask;
```
here and elsewhere, it often helps readability to make it clear what are the (non-static) data members of a class, by using some naming convention, like prefixing with `m_` or suffixing with `_` (prefixing with `_` can be problematic, as identifiers starting with `_` followed by an uppercase letter are reserved to the compiler)

stbergmann commented on makefile:
```diff
@@ -0,0 +1,19 @@
+
+clean:
+	rm -f $(OBJ_DIR)/*.o $(TARGET)
+
```
This lacks dependencies so that code is recompiled when only a `*.hpp` include file changes.  For a simple case like this, at a minimum you could specify the dependencies (incl. indirect ones, where one include file includes another one) explicitly like
```
$(OBJ_DIR)/bmp.o: $(SRC_DIR)/bmp.hpp
$(OBJ_DIR)/main.o: $(SRC_DIR)/bmp.hpp $(SRC_DIR)/pixel.hpp $(SRC_DIR)/pixelbasher.hpp
$(OBJ_DIR)/pixelbasher.o: $(SRC_DIR)/bmp.hpp $(SRC_DIR)/pixel.hpp $(SRC_DIR)/pixelbasher.hpp
```
or look into ways to do that more automatically (like <https://www.gnu.org/software/make/manual/make.html#Automatic-Prerequisites> or <https://www.gnu.org/software/automake/>)

stbergmann commented on src/bmp.cpp:
```diff
@@ -0,0 +1,339 @@
+		throw std::runtime_error(std::string("Can't open the BMP file: ") + filename);
+	}
+
+	input.read((char *)&file_header, sizeof(file_header)); // read file header data into struct
```
here and elsewhere: instead of using C-style casts (like `(char *)&file_header`, which can do all kinds of scary conversions "under the hood"), it's always better to be explicit with the various C++ `const_cast`, `static_cast`, `reinterpret_cast`  (in this case, you'd need a `reinterpret_cast<char *>(&file_header)`)

stbergmann commented on src/bmp.cpp:
```diff
@@ -0,0 +1,339 @@
+
+	input.seekg(file_header.offset_data, input.beg);
+
+	info_header.size = sizeof(BMPInfoHeader) + sizeof(BMPColourHeader); // we know that its a 32-bit BMP
```
one would assume that `info_header.size` is already set when `info_header` is read in, so if it is necessary to set it here for some reason, would be good to explain that with a (more explanatory) comment

stbergmann commented on src/bmp.cpp:
```diff
@@ -0,0 +1,339 @@
+	this->blurred_edge_mask = blur_edge_mask(*this, edge_mask);
+}
+
+BMP::BMP(const BMP &other)
```
you don't need to declare a copy constructor that explicitly does what the implicitly defined one would do anyway

stbergmann commented on src/bmp.hpp:
```diff
@@ -0,0 +1,107 @@
+	int32_t y_per_meter{0};
+	uint32_t colours_used{0};
+	uint32_t colours_important{0};
+};
```
for the members of `BMPFileHeader` and `BMPINfoHeader`, I would not specify default initializers---those values shall always be read from the given bmp file (IIUC?)

stbergmann commented on src/bmp.hpp:
```diff
@@ -0,0 +1,107 @@
+	uint32_t alpha_mask{0xff000000};
+	uint32_t colour_space{0x73524742}; // sRGB
+	uint32_t unused[16]{0};
+};
```
If I see correctly, `BMPColourHeader` is only ever used in `BMP::write` and `BMP::write_side_by_side`, to write its (hard-coded) content out to a file, right?  Which would mean we don't need a `BMP::colour_header` data member at all, and can just have a single global `static BMPColourHeader const colour_header;` in `bmp.jcpp` (and the definition of `BMPColourHeader` can then also be moved to `bmp.cpp`).

stbergmann commented on src/bmp.hpp:
```diff
@@ -0,0 +1,107 @@
+	std::vector<bool> sobel_edges(const BMP &image, int threshold = 40);
+	std::vector<bool> blur_edge_mask(const BMP &image, const std::vector<bool> &edge_map, int radius = 2);
+	std::array<int, 2> get_sobel_gradients(int y, int x, const std::vector<uint8_t> &data, int width, int pixel_stride);
+	void blur_pixels(int x, int y, int radius, int width, int height, std::vector<bool> &mask);
```
this can be a static member function (or could even be a plain function local to `bmp.cpp`?)

stbergmann commented on src/bmp.hpp:
```diff
@@ -0,0 +1,107 @@
+	int get_non_background_pixel_count(int background_value) const;
+
+	std::vector<bool> sobel_edges(const BMP &image, int threshold = 40);
+	std::vector<bool> blur_edge_mask(const BMP &image, const std::vector<bool> &edge_map, int radius = 2);
```
see above, regarding the `image` parameter

stbergmann commented on src/bmp.cpp:
```diff
@@ -0,0 +1,339 @@
+	non_background_count = get_non_background_pixel_count(background_value);
+
+	std::vector<bool> edge_mask = sobel_edges(*this);
+	this->blurred_edge_mask = blur_edge_mask(*this, edge_mask);
```
drop the `this->` from `this->blurred_edge_mask`

stbergmann commented on src/bmp.cpp:
```diff
@@ -0,0 +1,339 @@
+
+		if (padding_size > 0)
+		{
+			input.read((char *)padding.data(), padding_size); // read in the padding so next read is correct
```
instead of reading the padding (and storing it into an otherwise unused `vector`), just skip it instead with `input.seekg`?

stbergmann commented on src/bmp.cpp:
```diff
@@ -0,0 +1,339 @@
+		{
+			std::array<int, 2> gradients = get_sobel_gradients(y, x, data, width, pixel_stride);
+			int g_x = gradients[0];
+			int g_y = gradients[1];
```
you can directly use a structured binding here,
```
auto [g_x, g_y] = get_sobel_gradients(...);
```

stbergmann commented on src/bmp.cpp:
```diff
@@ -0,0 +1,339 @@
+	}
+}
+
+void BMP::write_side_by_side(BMP &diff, BMP &base, BMP &target, const char *filename)
```
the `diff`, `base`, `target` parameters can be of `const` reference type

ElliotFreebairn committed:
Fix: sorting headers lexicographically

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Refactor: dropping get functions for BMP headers

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Refactor:  removing parameters & optimising

sobel_edges and blur_ege_mask now belong to the object, and use this-> rather than taking in a BMP.

Introduced template variables for threshold and radius, as it makes sense for them to be evaluated at compile time for performance

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Refactor: adding prefix _m for member variables

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Fix: recompiles when *.hpp files are changed

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Refactor: removing redundant this-> expressions

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Refactor: swapping out c style casts to c++ casts

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Refactor: stop reading the padding from BMP's

The write functions still need to write the padding otherwise there will be errors.

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Refactor: using structured binding for
g_x and g_y

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Refactor: removing copy constructor

Compiler automatically creates a copy constructor, when needed, so no need to manually write one

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Refactor: making more objects (BMP and Pixel) const

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Refactor: removing unecessary code

The file  and info header data is already read in, so no need to double assign its values.

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Refactor: moving colour header out of header

BMPColourHeader only used in writing to BMP's and the values are hardcoded, so it's better to have a single global static colour_header defined once.

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

pedropintosilva commented:
@ElliotFreebairn could you please rebase this so all the commits are signed off , the old ones are missing that. thanks

ElliotFreebairn commented:
> @ElliotFreebairn could you please rebase this so all the commits are signed off , the old ones are missing that. thanks

@pedropintosilva All done now. 
Also shortened all the old commits to fit in the oneline command

stbergmann commented on makefile:
```diff
@@ -0,0 +1,20 @@
+-include $(OBJS:.o=.d)
+
+clean:
+	rm -f $(OBJ_DIR)/*.o $(TARGET)
```
you'd also want to clean the generated `$(OBJ_DIR)/*.d` files (or just `rm -fr $(OBJ_DIR)` completely, which is probably even better)

stbergmann commented on src/pixel.hpp:
```diff
@@ -0,0 +1,56 @@
+
+struct Pixel
+{
+	uint8_t blue{0};
```
you need to `#include <cstdint>` for this (and, strictly speaking, here and elsewhere: `<cstdint>` is guaranteed to provide `std::uint8_t` etc., but whether it also provides it as top-level `uint8_t` etc. is implementation defined, so better prefix all those std integer types with `std::` throughout)

stbergmann commented on .gitignore:
```diff
@@ -360,4 +360,19 @@ MigrationBackup/
+.vscode
+
+# Ignore build artifacts
+*.o
```
add `*.d`

stbergmann commented on src/bmp.cpp:
```diff
@@ -0,0 +1,334 @@
+    uint32_t unused[16]{0};
+};
+
+static BMPColourHeader colour_header = {
```
this can be `const`

stbergmann replied:
this should be `const` again; just in the places where you use it in a `reinterpret_cast`, those need to be `reinterpret_cast<char const *>(...)`

stbergmann commented on src/bmp.cpp:
```diff
@@ -0,0 +1,334 @@
+		throw std::runtime_error(std::string("Can't open the BMP file: ") + filename);
+	}
+
+	input.read(reinterpret_cast<char *>(&m_file_header), sizeof(m_file_header)); // read file header data into struct
```
Meaningfully reading the (little-endian) BMP data from the file into `m_file_header` this way only works on little endian architectures.  If you don't want to generalize the code (which might be overkill, for this app that will probably only ever be used on little endian machines?), you could document that it only works for little endian.  With C++20 (and I don't see a good reason you wouldn't want to use that?) you can use `#include <bit>` and `static_assert(std::endian::native == std::endian::little, "this code only works for little endian");`.

stbergmann replied:
also, better follow each `input.read` with an `if (!input) { throw std::runtime_error(...); }` check, to make sure we're not using uninitialized random data.  (you could also generally use `input.exceptions(std::ifstream::failbit);` to automatically have exceptions thrown whenever `input` goes into the fail state, but the resulting exception messages are probably not that nice and informative, when you want to print them at the bottom of `main`)

(admittedly, this is a bit of a back and forth, with me first suggesting to drop the default initializers from the `BMPFileHeader` and `BMPInfoHeader` structs, and now suggesting this mitigation of read failures, but I think it's overall cleaner this way)

stbergmann commented on src/bmp.cpp:
```diff
@@ -0,0 +1,334 @@
+    uint32_t blue_mask;
+    uint32_t alpha_mask;
+    uint32_t colour_space;
+    uint32_t unused[16]{0};
```
the default initializer `{0}` is unnecessary here, as the only use of `BMPColourHeader` below explicitly gives an initializer anyway

stbergmann commented on src/bmp.cpp:
```diff
@@ -0,0 +1,334 @@
+    0x000000ff,
+    0xff000000,
+    0x73524742,
+    {0}
```
I'd use either `{}` or `{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0}`, but not the confusing `{0}` (which explicitly initializes the first element of the array to `0` and implicitly initializes each of the remaining elements with `{}`, i.e., `0` as well).

stbergmann commented on src/main.cpp:
```diff
@@ -0,0 +1,254 @@
+	}
+	else
+	{
+		throw std::runtime_error("Incorrect usage for " + option_name + ": " + value + " should be true of flase");
```
typo `flase` -> `false`

stbergmann commented on src/main.cpp:
```diff
@@ -0,0 +1,254 @@
+
+void parse_flags(int &arg_index, char *argv[], ParsedArguments &args)
+{
+	std::string minor_differences = argv[arg_index - 1];
```
I'd move reading `argv[arg_index - 1]` into `parse_flag`

stbergmann commented on src/main.cpp:
```diff
@@ -0,0 +1,254 @@
+}
+
+void parse_directories(int &arg_index, char *argv[], ParsedArguments &args)
+{
```
can't hurt to do an `assert(arg_index >= 5` here, to proof that all the following accesses into `argv` are valid

ElliotFreebairn committed:
Fix: Remove imagemagick/wand

The script can now run on the windows server as all references to wand are gone + the path to the program is properly set

Fix: colour_header not copied correctly

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

stbergmann commented on src/bmp.cpp:
```diff
@@ -0,0 +1,334 @@
+
+	BMPFileHeader file_header = diff.m_file_header;
+	BMPInfoHeader info_header = diff.m_info_header;
+	BMPColourHeader colour_header = colour_header;
```
this is broken (it initializes the local variable `colour_header` with the local variable `coulour_header` itself; i.e., "copies" uninitialized values), and unnecessary---you can directly use the global `colour_header` variable in the code below

stbergmann commented on src/pixelbasher.cpp:
```diff
@@ -0,0 +1,166 @@
+		return {0, 255, 0, 255};
+		break;
+	default:
+		throw std::runtime_error("Invalid colour enum passed " + colour);
```
To add the number representation of an integer value to a string, you need to use something like `std::to_string`  (the above does something completely different ;)

(Also, this default branch is more of a "this cannot happen" logic error than a runtime error (which would warrant throwing an exception).  Better to drop the default branch and instead end the function with `assert(false);` or C++23 `std::unreachable();`, to indicate that it would be a logic error if we would ever reach the bottom of the function and not return anything from a function that must return something.)

stbergmann commented on src/pixelbasher.cpp:
```diff
@@ -0,0 +1,166 @@
+	{
+	case Colour::YELLOW:
+		return {0, 197, 255, 255};
+		break;
```
the `break` after `return` is superfluous

stbergmann commented on src/pixelbasher.cpp:
```diff
@@ -0,0 +1,166 @@
+{
+	int min_width = std::min(original.get_width(), target.get_width());
+	int min_height = std::min(original.get_height(), target.get_height());
+	int pixel_stride = 4;
```
`pixel_stride` can be a global `constexpr`, shared between this function and `PixelBasher::compare_regressions`  (and potentially shared with even more code, see the below suggestion to use `std::array<uint8_t, pixel_stride>` instead of `std::vectoryuint8_t>` throughout)

stbergmann commented on src/pixelbasher.cpp:
```diff
@@ -0,0 +1,166 @@
+			Pixel target_pixel = Pixel::get_pixel(target_data, target_index);
+			bool near_edge = intersection_mask[mask_index];
+
+			std::vector<uint8_t> bgra = compare_pixels(original_pixel, target_pixel, diff, near_edge, enable_minor_differences);
```
The following code uniformly operates on the individual members of `bgra`, so it is arguably better to use a `vector` than to use `Pixel` with its individual `blue`, `green`, `red`, `alpha` members.  But `vector` is rather "fat" and "expensive" (and doesn't say that it always contains four elements), so a `std::array<uint8_t, pixel_stride>` would arguably be an even better type here.

stbergmann reviewed (commented):
here's another batch of comments; let me know if you want to discuss any of that in a video chat, otherwise I'd say we can continue by exchanging comments here for the time being?

ElliotFreebairn commented:
> here's another batch of comments; let me know if you want to discuss any of that in a video chat, otherwise I'd say we can continue by exchanging comments here for the time being?

Thank you

Im good with exchanging comments for the meantime. If I'm unsure/stuck I'll message you

ElliotFreebairn committed:
Remove dependency files in make clean

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Prefixing int/uint values with std::

Also including <cstindt> in header files where needed.

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Make header const + add .d files to gitignore

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Throw error if user is not running little-endian

Updated the makefile to run c++ 20 which includes std::endian

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Remove useless initalizer for colour header

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Read option values inside of parse_flag

Instead of manually wiriting argv[arg_index - 1] for each flag, do it once in parse_flag

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Adding checks when  parsing directories and options

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Add extra checks when reading in BMP

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Removing useless returns and default result

Instead of throwing an error, assert(false) to indicate that it's a logic error.

Essentially colour_pixel has to return a value, so if no value is returned, that's a logic error.

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Improving performance via pixel access/edits

Instead of pixels being stored as a Pixel struct, we create a array of 4 bytes, where each byte represents B,G,R,A

Then the Pixel struct contains static methods to compare those arrays(pixels)

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Fix: can set density, resize and max_pages

The density option was not applying to the conversion previously because of the background white option

Now the user can set the maximum amount of pages they want to convert, how much the image should be scaled down and the dpi

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Improving sensitivity to subtle pixel changes

A new colour is added, dark yellow.

It indicates that a pixel is located on an edge and is significantly different

The dark yellow pixel may be a false positive.

However it has a much higher chance of being significantly different.

Also reduced the radius of pixels blurred near edges/text

 This reduces the amount of pixels ignored for differences.

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Feat: automatically labeling side by side images

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Feat: adding the no_save_overlay option

By default the BMP's are saved.

There is an option --no_save_overlay, that allows BMP's not be saved, only if the red count in the current COOL page exceeds the previous COOL page

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Feat: add vertical edge detection

It is only true for a pixel to be a vertical edge, if its part of a longer continous run of vertical edges

This is to avoid highlighting the vertical edges of text, horizontal lines etc...

Then if a pixel is considered to be different, but a part of a vertical edge, it is highlighted yellow

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Change when  and what images are outputted

Now this program mimicks excatly what justin wrote with the old imagemagick code (plus more features)

There are different options such as no_save_overlay, image_dump which change what is outputed at the end.

When no_save_overlay is enabled, only pages which have higher red count than the previous version are outputted

Whne image dump is enabled, the normal images are produced plus the side by side comparisons and the original BMP's used to diff

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Recording previous image data in csv + clean up

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Fix: side by side comparison not being written

This was due to the fact it would only write if lo/ms preivous existed, which is incorrect as logically it doesn't depend on it

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Make stamps larger and red

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

stbergmann commented on src/pixel.hpp:
```diff
@@ -0,0 +1,47 @@
+	std::uint8_t red{0};
+	std::uint8_t alpha{0};
+
+	static std::array<std::uint8_t, pixel_stride> get_vector(const std::uint8_t* src_row)
```
* it might make sense to introduce a top-level typedef for `std::array<std::uint(-t, pixel_sttide>` and use it thoughout; the modern C++ way of defining a typedef is
```
using PixelValues = std::array<std::uint8_t, pixel_stride>;
```
(and some places still use `std::array<std::uint8_t, 4>` instead of `std::array<std::uint8_t, pixel_stride>`)

* `get_vector` is a bit of a misnomer now

stbergmann commented on src/pixel.hpp:
```diff
@@ -0,0 +1,47 @@
+	}
+
+	// Compare this pixel with another pixel and return true if they differ
+	static bool differs_from(std::array<std::uint8_t, 4> original, std::array<std::uint8_t, 4> target, bool near_edge, int threshold = 40)
```
(In general, when passing potentially large data into a function, it is better to pass by const lvalue reference rather than by value.  In this particular case, `std::array<std::uint8_t, pixel_stride>` turns out to be a rather small 4 bytes in size, so you might just as well keep it this way.)

stbergmann commented on src/bmp.cpp:
```diff
@@ -0,0 +1,440 @@
+};
+
+
+BMP::BMP(const char *filename, std::string basename)
```
the `basename` parameter is unused

stbergmann commented on src/bmp.cpp:
```diff
@@ -0,0 +1,440 @@
+		throw std::runtime_error("Cannot open/create the file to write");
+	}
+
+	BMPColourHeader local_colour_header = ::colour_header;
```
Here and elsewhere: you don't need to first copy the global `colour_header` into a local variable, you can directly use the global variable in the code below.  (And the `::` prefix to access a global name, absent any name clash that would require it, might be considered "noise" that can just as well be dropped.)

stbergmann commented on src/bmp.cpp:
```diff
@@ -0,0 +1,440 @@
+	std::string cool_location = stamp_location + "/cool.bmp";
+
+	BMP diff_stamp(diff_location.c_str(), "diff");
+    BMP ms_office_stamp(ms_office_location.c_str(), "office");
```
There's an odd mix of using tabs and spaces for indentation here.  (My personal recommendation would be to never use tabs in source code, other than where it is syntactically necessary, i.e., at the start of recipe lines in makefiles), because that overall gives a better user experience---it is independent of how tools used to display source code are actually displaying tabs if you don't use them in the first place.)

ElliotFreebairn committed:
Using a top-level typedef for bgra/pixel arrays

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Removing unused basename in BMP constructor

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Use colour_header directly instead of copying

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Use 4 spaces for indentation instead of tabs

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Cleaning up and removing comments

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Fix: converted images not deleted if an error is raised

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

stbergmann reviewed (commented):
there's still places that use `std::array<std::uint8_t, 4>` and could use `PixelValues` instead

ElliotFreebairn committed:
Replacing missed std::arrays with PixelValues

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Add a small test case with make check

There is a converted and download folder containing the input pdf's/doc and then an expected folder which contains the reference output which is checked against the output prdouced by the program

Also fixed a bug where the original image was stamped and then diffed, rather than it being stamped at the end

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn commented:
@stbergmann I've created a small test/check case. Am I on the right track with that?

stbergmann commented on makefile:
```diff
@@ -0,0 +1,33 @@
+	echo "Test passed" || echo "Test failed"
+
+clean:
+	rm -fr $(OBJ_DIR) $(TARGET)
```
after a `make check && make clean && git ls-files -o --directory`, see what's still lying around and should be removed here

stbergmann commented on makefile:
```diff
@@ -0,0 +1,33 @@
+-include $(OBJS:.o=.d)
+
+check:
+	@mkdir -p test/output ./converted/import ./converted/import-compare  ./converted/export ./converted/export-compare
```
I'd never use `@` prefixes in makefile recipes; if people want to have short output, they can run `make -s`; and if they want to see every command that gets executed (e.g., to debug something wrong with a makefile), they want to see *every* command really.

stbergmann commented on makefile:
```diff
@@ -0,0 +1,33 @@
+	@mkdir -p test/output ./converted/import ./converted/import-compare  ./converted/export ./converted/export-compare
+	@rm -rf converted/import converted/import-compare converted/export converted/export-compare
+	@echo "Running test with python script"
+	@python3 diff-pdf-page-statistics.py \
```
it took me by surprise that `diff-pdf-page-statistics.py` internally invokes `make`; I'd drop that (and have that tool only be usable after this project has been "installed", by running `make`), and make this `check` target depend on `$(TARGET)`.

Also, the `check` target can be `.PHONY` (as it doesn't actually produce a file named `check`; see the GNU Make documentation for details of phony targets if you want to learn more)

stbergmann commented on makefile:
```diff
@@ -0,0 +1,33 @@
+		--resolution=75
+	@cmp --silent ./converted/import/doc/fdo69695-1.doc_import-1.bmp ./converted/expected/fdo69695-1.doc_import-1.bmp && \
+	cmp --silent ./converted/export/doc/fdo69695-1.doc_export-1.bmp ./converted/expected/fdo69695-1.doc_export-1.bmp && \
+	echo "Test passed" || echo "Test failed"
```
I'd turn this into two individual recipe lines invoking `cmp` (not linked by `GG`), and drop their `--silent` switches (so that if they fail, it is obvious which of them failed, and why), and drop the trailing two `echo`s---`make` already tells you whether making the target succeeds (i.e., `make` itself remains silent) or failed (i.e., `make` prints some `make: *** [makefile:24: check] Error 1`)

Btw, for me this then always fails with
```
python3 diff-pdf-page-statistics.py \
	--base_file=fdo69695-1.doc \
	--max_page=1 \
	--minor_differences=false \
	--resolution=75
Processing:  fdo69695-1.doc
cmp ./converted/import/doc/fdo69695-1.doc_import-1.bmp ./converted/expected/fdo69695-1.doc_import-1.bmp
./converted/import/doc/fdo69695-1.doc_import-1.bmp ./converted/expected/fdo69695-1.doc_import-1.bmp differ: byte 77411, line 1
make: *** [makefile:24: check] Error 1
```

ElliotFreebairn replied:
Hmm that's weird.
This is my output:
```
python3 diff-pdf-page-statistics.py \
	--base_file=fdo69695-1.doc \
	--max_page=1 \
	--minor_differences=false \
	--resolution=75
Processing:  fdo69695-1.doc
cmp  ./converted/import/doc/fdo69695-1.doc_import-1.bmp ./converted/expected/fdo69695-1.doc_import-1.bmp
cmp  ./converted/export/doc/fdo69695-1.doc_export-1.bmp ./converted/expected/fdo69695-1.doc_export-1.bmp
```

For some reason both of you're comparing is being done on import files not the export ones. Maybe try again with fixes, and let me know if it's still breaking.

stbergmann replied:
Given that `diff-pdf-page-statistics.py` calls `magick` first to convert from pdf to bmp, maybe what's the issue here is different versions of `magick` generating slightly different bmps (my version is `ImageMagick-7.1.1.47-1.fc42.x86_64`).  That would unfortunately mean that we can't (easily) have an end-to-end test that calls `diff-pdf-page-statistics.py`, but would rather need to do a test directly on checked-in bmp files and calling `pixelbasher` directly?

(Also, I noticed that the checked-in pdfs like `converted/doc/fdo69695-1.doc_import.pdf` have their (Unix) execute bits set, which can be removed. ---and which is moot if they need to be replaced with bmps anyway)

ElliotFreebairn replied:
Oh okay that would make sense. I'll get the right BMP's put in source and switch from the python script to pixelbasher. Thank you

ElliotFreebairn replied:
@stbergmann I've pushed a commit hopefully solving this issue

stbergmann reviewed (commented):
(In general, it is a bit confusing when a commit like `a4fe31478103305c61c691d9127cb2f0bf380e98` "Add a small test case with make check" also includes a completely unrelated `colour_header` `const` fix.  But I'm mostly not looking at your individual commits but rather just at the resulting `HEAD`, anyway.)

ElliotFreebairn committed:
Removing test images and CSV's in make clean

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Removing @ prefix for better debugging

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Remove make from python and add .PHONY targets

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Drop --silent and && between compare commands

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Feat: a new csv is written to containg pdf anomalies

Initially, the program will convert all pages to images, so we can record the number of pages for each pdf.

If there a mismtach in number of pages, that information is written to a separate csv file

After that, the pages are trimmed so that there are only max_pages, or the minimum number of pages from one of the PDF'S in the set.

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Fix: crashing in write_side_by_side

If any of three images had a different dimensions, the program returned a `munmap_chunk(): invalid pointer`

Now we check if there are any mismatches, display that to the user and skip that image write

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Removing unused copies of image data

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Refactor: rename converted bmps

The exported/ms_conv bmp's were called import_mso which is a bit misleading and hard to differentiate

Changed them to export-*.bmp to better differentiate

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Fix: make check now uses input bmp's directly

Due to issues with image magick versions, there is now a input folder which contains converterd BMP's

They are then compared byte by byte against the BMP's located inside expected/

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Adjust edge detection   + display filtered images

The sobel edge detection threshold was increased so that slightly non-white pixels weren't picked up as edges

Also added the ability to display the effects of edge detection, vertical edges and filtering vertical edges for debugging purposes

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Improve vertical edge filter by lowering threshold

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Fix: conversion to BMP too long and poor quality

Each image was being resized to 75% and this led to horibbly long conversion times

But even worse, it reduced the quality of image so much, text and drawings because transparent/white, leading to false positives

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Better thresholding based off background value

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

ElliotFreebairn committed:
Updating test images after removing resize option

Signed-off-by: Elliot Freebairn <elliot.freebairn@collabora.com>

stbergmann commented:
LGTM
